File: examples/java/02_java_main_api/src/main/java/io/touca/examples/main/StudentsTest.java
// Copyright 2021 Touca, Inc. Subject to Apache-2.0 License.

package io.touca.examples.main;

import io.touca.Touca;
import io.touca.TypeAdapterContext;
import java.time.LocalDate;

public final class StudentsTest {

  @Touca.Workflow
  public void findStudent(final String username) {
    Touca.startTimer("find_student");
    Student student = Students.findStudent(username);
    Touca.stopTimer("find_student");
    Touca.assume("username", student.username);
    Touca.check("fullname", student.fullname);
    Touca.check("birth_date", student.dob);
    Touca.check("gpa", student.gpa);
    Touca.check("pass", student.gpa < 3.9);
    Touca.addMetric("external_source", 1500);
  }

  public static void main(String[] args) {
    Touca.addTypeAdapter(LocalDate.class, (x) -> {
      TypeAdapterContext object = new TypeAdapterContext();
      object.add("day", x.getDayOfMonth());
      object.add("month", x.getMonthValue());
      object.add("year", x.getYear());
      return object;
    });
    Touca.run(StudentsTest.class, args);
  }

}

File: examples/cpp/04_cpp_external_input/students_test.cpp
// Copyright 2021 Touca, Inc. Subject to Apache-2.0 License.

#include "students.hpp"

#include <iostream>
#include <thread>

#include "cxxopts.hpp"
#include "touca/core/filesystem.hpp"
#include "touca/touca.hpp"

struct CustomConfigurationOutput {
  touca::filesystem::path datasets_dir;
  std::vector<std::string> testcases;
};

CustomConfigurationOutput custom_configuration(int argc, char* argv[]) {
  CustomConfigurationOutput out;
  cxxopts::Options opts{""};
  opts.add_options()("datasets-dir", "path to datasets directory",
                     cxxopts::value<std::string>());
  opts.allow_unrecognised_options();
  const auto& result = opts.parse(argc, argv);
  if (!result.count("datasets-dir")) {
    throw std::invalid_argument(
        "required configuration option \"datasets-dir\" is missing");
  }
  out.datasets_dir = result["datasets-dir"].as<std::string>();
  if (!touca::filesystem::is_directory(out.datasets_dir)) {
    throw std::invalid_argument("datasets directory \"" +
                                out.datasets_dir.string() +
                                "\" does not exist");
  }
  for (const auto& it :
       touca::filesystem::directory_iterator(out.datasets_dir)) {
    if (!touca::filesystem::is_regular_file(it.path().string())) {
      continue;
    }
    out.testcases.push_back(it.path().stem().string());
  }
  return out;
}

void custom_execution(const touca::filesystem::path& caseFile) {
  const auto& student = find_student(caseFile.string());
  touca::assume("username", student.username);
  touca::check("fullname", student.fullname);
  touca::check("birth_date", student.dob);
  touca::check("gpa", calculate_gpa(student.courses));
  custom_function_1(student);
  std::thread t(custom_function_2, student);
  t.join();
  touca::start_timer("func3");
  custom_function_3(student);
  touca::stop_timer("func3");
  touca::add_metric("external", 10);
}

int main(int argc, char* argv[]) {
  auto out = custom_configuration(argc, argv);
  std::cout << out.testcases.size() << std::endl;
  touca::workflow(
      "external_input",
      [&out](const std::string& testcase) {
        custom_execution(out.datasets_dir / (testcase + ".json"));
      },
      [&out](touca::WorkflowOptions& x) { x.testcases = out.testcases; });
  return touca::run(argc, argv);
}

template <>
struct touca::serializer<Date> {
  data_point serialize(const Date& value) {
    return object("Date")
        .add("year", value.year)
        .add("month", value.month)
        .add("day", value.day);
  }
};
